XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+by handles to perform work. Such as accepting a new connection or writing data
+to disk.
+
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// Return the id of the current execution context. Useful for tracking state
+// and retrieving the handle of the current parent without needing to use an
+// AsyncHook().
+const id = async_wrap.currentId();
+
+// Generate new uid for user API.
+const id = async_wrap.newUid();
+
+// Create a new instance and add each callback to the corresponding hook.
+var asyncHook = new async_wrap.AsyncHook({ init, before, after, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+// Enable hooks for the current synchronous execution scope. This will ensure
+// the hooks are not in effect in case of multiple returns, or if an exception
+// is thrown.
+asyncHook.syncScope();
+
+// Unlike disable(), prevent any hook callback from firing again in the future.
+// Future hooks can be added by running syncScope()/enable() again.
+// TODO: Investigation is needed to know whether this is possible.
+asyncHook.remove();
+
+// init() is called during object construction. The handle may not have
+// completed construction when this callback runs. So all fields of the
+// handle referenced by "id" may not have been populated.
+function init(id, type, parentId, handle) { }
+
+// before() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWrap), and should be called exactly 1
+// time for requests (e.g. FSReqWrap).
+function before(id) { }
+
+// after() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw.
+function after(id) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPParser where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after after() has
+// completed.
+function destroy(id) { }
+```
+
+
+### `async_wrap`
+
+The object returned from `require('async_wrap')`.
+
+
+#### `async_wrap.currentId()`
+
+Return the id of the current execution context. Useful to track the
+asynchronous execution chain. It can also be used to propagate state without
+needing to use the `AsyncHook` constructor. For example:
+
+```js
+const map = new Map();
+
+net.createServer((c) => {
+  // TODO: In order for this to be useful it needs to be possible to get the
```

Comment:
That's a line item for and API that doesn't exist yet that I would like to figure out. Which is, to get the uid of any given handle. For example, the net API has some conflicting parts I'm reasoning with.
```js
// 1) We can't get the id of the server handle here b/c no handle
//    is actually created until .listen() is called.
const server = net.createServer((c) => {
  // 3) It's noted at the bottom of the EP that this id would be
  //    manually made the connection id, instead of the server id.
  //    Which it technically should be. This is because there's no
  //    other way to retrieve the connection's id until after an event
  //    on it has fired.

  // 4) What would be most convenient is if the user could call
  //    server.asyncId(), instead of needing to retrieve it in .listen(),
  //    cache it somewhere then retrieve it here.
});

server.listen(8080, () => {
  // 2) This is the soonest the id of the server can be retrieved.
  //    Which then would need to be cached for later when the
  //    'connection' callback is called.
});
```
So this `TODO` is a note to myself to figure out a way to support something like `server.asyncId()` _and_ the embedder API.